modules/internal/String.chpl

```diff
@@ -1337,9 +1470,18 @@ module String {
       var result: string = this.toLower();
       if result.isEmptyString() then return result;
 
-      var b = result.buff[0];
-      if byte_isLower(b) {
-        result.buff[0] = b - 0x20;
+      var codepoint: int(32);
+      var nbytes: c_int;
+      var multibytes = result.buff: c_string;
+      var maxbytes = result.len: ssize_t;
+      if maxbytes > 4 then
+        maxbytes = 4;
+      qio_decode_char_buf(codepoint, nbytes, multibytes, maxbytes);
+      var upCodepoint = codepoint_toUpper(codepoint);
+      if upCodepoint != codepoint {
+        // This assumes that the upper and lower case version of a
```

Comment:
Just FYI this doesn't work generally, see http://unicode.org/faq/casemap_charprop.html SpecialCasing.txt linked from there has all the exceptions ( ftp://ftp.unicode.org/Public/UCD/latest/ucd/SpecialCasing.txt )

Reply:
Could you elaborate?  I don't understand what doesn't work.

Reply:
Case operations on Unicode strings may cause the length of the string to change, and there is not necessarily a 1:1 mapping of codepoints. Like http://unicode.org/faq/casemap_charprop.html#4 mentions, title casing is a bit special as well, and doing it naively will not always be correct.

Reply:
Ah, I think I see what you're saying.  One upper case codepoint might map to a sequence of lower case codepoints, for example.  I do not attempt to handle that case.  This change uses the underlying C library, which has no way to handle one-to-many mappings.

Reply:
Yup, I was just pointing out that this not the correct behavior for Unicode strings. I'd strongly recommend looking into using something like https://github.com/JuliaStrings/utf8proc for the heavy lifting in these functions.